src/FSharpPlus/Data/Seq.fs

```diff
@@ -16,73 +16,885 @@ module Seq =
+                    let current = ref Option<'U>.None
+                    { new IEnumeratorM<'``Monad<bool>``, 'U>  with
+                        member _.Current =
+                            match !current with
+                            | Some c -> c
+                            | None -> invalidOp "Enumeration has not started. Call MoveNext."                            
+                        member x.MoveNext () =
+                            monad' {
+                                match !state with
+                                    | CollectState.NotStarted f ->
+                                        return! (
+                                            let e1 = (f :> IEnumerableM<'``Monad<bool>``, ('T -> 'U)>).GetEnumerator ()
+                                            state := CollectState.HaveInputEnumerator e1
+                                            x.MoveNext ())
+                                    | CollectState.HaveInputEnumerator e1 ->
+                                        let! res1 = e1.MoveNext ()
+                                        return! (
+                                            if res1 then
+                                                let e2 = (x1 :> IEnumerableM<'``Monad<bool>``, 'T>).GetEnumerator ()
+                                                state := CollectState.HaveInnerEnumerator (e1, e2)
+                                            else
+                                                x.Dispose ()
+                                            x.MoveNext () )
+                                    | CollectState.HaveInnerEnumerator (e1, e2) ->
+                                        let! (res2: bool) = e2.MoveNext ()
+                                        if res2 then
+                                            current := Some (e1.Current e2.Current)
+                                            return res2
+                                        else
+                                            state := CollectState.HaveInputEnumerator e1
+                                            dispose e2
+                                            return! x.MoveNext ()
+                                    | _ ->
+                                        return false }
+                        member _.Dispose () =
+                            match !state with
+                            | CollectState.HaveInputEnumerator e1 ->
+                                state := CollectState.Finished
+                                dispose e1
+                            | CollectState.HaveInnerEnumerator (e1, e2) ->
+                                state := CollectState.Finished
+                                dispose e2
+                                dispose e1
+                            | _ -> () } }
+
+    let inline lift2<'T1, 'T2, 'U, .. > (f: 'T1 -> 'T2 -> 'U) (x1: SeqT<'``Monad<bool>``, 'T1>) (x2: SeqT<'``Monad<bool>``, 'T2>) : SeqT<'``Monad<bool>``, 'U> =
```

Comment:
Out of curiosity, what is `..` in the generic type declaration? Can't believe I've seen that before.

Reply:
That's a rarely used syntax, it means "infer the rest of the type parameters with their constraints". I wish `< .. >` was also allowed.

Reply:
Cool!